### ISSUE TITLE
`updateMany` fails on `data` array of existing records

### ISSUE THREAD

**Issue by daneden** [prisma/prisma#10382]
### Bug description

I’m not sure if I’m using `updateMany` incorrectly but I’m experiencing an odd bug.

I have an array of model records retrieved using `findMany`, updated via `Array.map`, and I’m attempting to then update the records using `updateMany`.

I get an error that complains about passing incorrect types to `data`, but don’t see how that’s possible. My hunch is that it’s actually a problem with using `updateMany` to update records with different values per record id—the documentation suggests `updateMany` is better suited to update many records with the _same_ values.

Am I using `updateMany` wrong, or is this unexpected behaviour? Thanks in advance!

### How to reproduce

1. Run an `updateMany` query against records with specific IDs and differing values per ID
2. See type error

### Expected behavior

Records should be updated to match the passed `data` array

### Prisma information

```prisma
model User {
  id Int @id @default(autoincrement())
  bestFriendId Int?
  bestFriend User? @relation(fields: [bestFriendId], references: [id])
}
```

```ts
const records = prisma.user.findMany()
const updatedRecords = records.map(user => {
  user.bestFriendId = findBestFriendId(user) // either an Int or null/undefined
  return user
}

await prisma.user.updateMany({ data: updatedRecords })
```

### Environment & setup

- OS: macOS
- Database: MySQL/Planetscale
- Node.js version: 16.13.0

### Prisma Version

```
3.5.0
```

**Comment by Andy671** [comment]
The same question ... Have you figured out how to use **updateMany** with different values per id?

**Comment by FLevent29** [comment]
I just ran into this problem as well. Looking at the types generated this isn't a bug, it's simply not supported. You have to run an update for each separately. So there's no `upsertMany` and `updateMany` is quite limited, maybe that's why there's no `upsertMany` yet in the first place.

**Comment by pantharshit00** [comment]
@daneden Can you please tell me what type error you are getting? I tried you example locally and it didn't throw any type error. Maybe share the `findBestFriendId` function as well.

**Comment by pantharshit00** [comment]
Yeah, `updateMany` will mutate the provided data where the condition in `where` is met. If you want to mutate multiple records with different data, then you will need to execute `update` method on each of the data point.

**Comment by marcelherd** [comment]
@pantharshit00 does that work properly?

I have an array of ids and would like to update one column for each of them:

```typescript
const matchIds = [ 8, 9, 10, 11, 12 ];

db.match.updateMany({
  where: {
    id: { in: matchIds },
  },
  data: {
    recordingUrl: videoUrl,
  },
});
```

I can see that the code ran, I can see that the matchIds exist in my (SQLite) database, but nothing happens. I set the environment variable `DEBUG=prisma*`. When I created the matches initially, I could see mutations being logged. But when the updateMany function runs, it doesn't log anything, no query, no mutation, nothing.

This is my schema:

```
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:./database.db"
}

model Match {
  id                Int      @id @default(autoincrement())
  playerCharacter   String
  playerRank        String
  opponent          String
  opponentCharacter String
  opponentRank      String
  stage             String
  playerRoundWins   Int
  opponentRoundWins Int
  roundWinsRequired Int
  recorded          Boolean  @default(false)
  recordingUrl      String?
  createdAt         DateTime @default(now())
  updatedAt         DateTime @updatedAt
}
```

Am I using it wrong or is it not supported for SQLite?

**Comment by heymartinadams** [comment]
A somewhat required feature in my view 🙂

Here’s a use-case: **encryption**. For encryptions to work, in addition to an encryption key, you usually also need to inject an **initialization vector** into the encryption to make the encryption truly randomized and unbreakable. A initialization vector is a random key that:
1. has to be _unique_ to each record (unlike the encryption key, which remains the same for all records);
2. has to be known for decryption; and
3. doesn’t have to be kept a secret (unlike the encryption key)

For that, a record’s `id` — it being a `uuid()` — works ideal as an initialization vector. Thus, an `updateMany` operation will require each record’s `id` _during_ the operation for an encryption to work.

### Pseudo code example using a callback.
```js
await prisma.todo.updateMany({
  where,
  data: record => ({
    ...data,
    ...(data?.text ? { text: encrypt(data.text, record.id) } : {}),
  }),
  select: { count: true }
})
```

**Comment by heymartinadams** [comment]
ps: this would also be an even better use case for `update()` since most encryption is updated individually (e.g. someone updating an individual record only).

**Comment by heymartinadams** [comment]
Adding this to the discussion: https://twitter.com/ruheni_alex/status/1592832451013128194

**Comment by njoshi22** [comment]
I'm having this issue in June of 2024 - was any bug, or update every released to address this?
